target_decisioning_engine/artifact_provider.py

```diff
@@ -0,0 +1,198 @@
+        self.polling_halted = False
+        self.polling_timer = None
+        self.artifact = None
+        self.subscriptions = {}
+        self.subscription_count = 0
+        self.last_response_etag = None
+        self.last_response_data = None
+        self.artifact_tracer = None
+
+    def _get_polling_interval(self):
+        """Get artifact polling interval"""
+        if self.config.polling_interval == 0:
+            return 0
+
+        return max(
+            MINIMUM_POLLING_INTERVAL,
+            self.config.polling_interval if is_number(self.config.polling_interval) else DEFAULT_POLLING_INTERVAL
+        )
+
+    def initialize(self):
+        """Initialize ArtifactProvider and fetch initial artifact"""
+        self.polling_interval = self._get_polling_interval()
+        self.artifact_location = self.config.artifact_location if is_string(self.config.artifact_location) else \
+            determine_artifact_location(self.config)
+        try:
+            self.artifact = self._get_initial_artifact()
+            # GA TODO - ArtifactTracer is a separate ticket
+            # self.artifact_tracer = ArtifactTracer(
+            #     self.artifact_location,
+            #     self.config.artifact_payload,
+            #     self.polling_interval,
+            #     self.polling_halted,
+            #     self.artifact
+            # )
+            # self.add_subscription(self.artifact_tracer_update)
+        finally:
+            self._schedule_next_update()
+
+    def _emit_new_artifact(self, artifact_payload, geo_context=None):
+        """Send events and notify subscribers of new artifact"""
+        if not geo_context:
+            geo_context = {}
+
+        self.event_emitter(ARTIFACT_DOWNLOAD_SUCCEEDED, {
+            "artifact_location": self.artifact_location,
+            "artifact_payload": artifact_payload
+        })
+
+        self.event_emitter(GEO_LOCATION_UPDATED, {
+            "geo_context": geo_context
+        })
+
+        for subscription_func in self.subscriptions.values():
+            subscription_func(artifact_payload)
+
+    def add_subscription(self, callback_func):
+        """Add event subscription"""
+        self.subscription_count += 1
+        self.subscriptions[self.subscription_count] = callback_func
+        return self.subscription_count
+
+    def remove_subscription(self, _id):
+        """Remove event subscription"""
+        try:
+            del self.subscriptions[_id]
+            self.subscription_count -= 1
+        except KeyError:
+            pass
+
+    def _fetch_and_schedule(self):
+        """Fetch artifact and schedule next polling"""
+        self.artifact = self._fetch_artifact(self.artifact_location)
+        self._schedule_next_update()
+
+    def _schedule_next_update(self):
+        """Schedule next artifact polling based on configured interval (in seconds)"""
+        if self.polling_interval == 0 or self.polling_halted:
+            return
+
```

Comment:
Do you need to call `self.polling_timer.cancel()` to clean up before creating a new `Timer`?

Reply:
Double checked docs and no, don't need to